Reimplement invocation of per-message Filter methods

robobario opened the pull request:
### Problem

Kroxy is vulnerable to a JVM performance problem ([JDK-8180450](https://bugs.openjdk.org/browse/JDK-8180450)). A filter that implements two or more specific-message Filter interfaces its secondary_super_cache (in its Klass instance within the JVM) would frequently be changing between those interfaces, with cache line interference result in poor performance on some hardware.

### Solution

Instead of calling the KrpcFilter defined methods on a Per-Message filter implementation we instead construct a Per-Message Invoker for that filter. The invoker determines all the Per-Message interfaces that are implemented at construction time. It tracks which request and response APIs need to be deserialized based on which interfaces are implemented. If a filter implements Per-Message API X, then we cast it and store it in a field of type X. Then at filtering time we can check if an ApiKey should be handled, and then delegate to the appropriate field.
    
We also enable the Per-Method interfaces to further specialize which messages should be deserialized by invoking their shouldDeserialize* methods if they match the incoming messages ApiKey.
    
Note that we are also laying the groundwork for alternative implementations of KrpcFilter. So `KrpcFilter` now has a method `KrpcFilter of(Object filter)` which will introspect the object. For now it will wrap it in a PerMessageInvoker if it implements any Per-Message interface. In future we can stop the Per-Message interfaces extending KrpcFilter and offer alternative interfaces for different use-cases (like intercepting all messages conveniently). This will decouple the internals of kroxy from the filter interfaces users implement to build out new filters.

robobario commented on kroxylicious/src/main/templates/Kproxy/Filter.ftl:
```diff
@@ -47,6 +51,10 @@ public interface ${filterClass} extends KrpcFilter {
      */
-    public void on${messageSpec.name}(${dataClass} ${msgType}, KrpcFilterContext context);
+<#if messageSpec.type?lower_case == 'request'>
+    public void on${messageSpec.name}(DecodedRequestFrame<${dataClass}> ${msgType}, KrpcFilterContext context);
```
I got the feeling `DecodedRequestFrame` might be the wrong abstraction level here. We could alternatively pass the `body` and `header` as separate parameters. We should also update the KrpcFilterContext methods to match depending on what we pick as mutating `frame.headers` and then calling `forwardRequest(body)` feels wrong.

k-wall replied:
I agree DRF feels wrong. I like the idea of separate header and body parameters, with the obligation that the implementor forwardRequest(header, body).

robobario commented on kroxylicious/src/main/java/io/kroxylicious/proxy/filter/KrpcFilter.java:
```diff
@@ -0,0 +1,70 @@
+ *
+ * <p>When implementing one or more of the {@code *RequestFilter} subinterfaces you need only implement
+ * the {@code on*Request} method(s), unless your filter can avoid deserialization in which case
+ * you can override {@link #shouldDeserializeRequest(ApiKeys, short)} as well.</p>
```
comment outdated, message-specific Filters are no longer KrpcFilters

robobario commented on kroxylicious/pom.xml:
```diff
@@ -217,7 +217,7 @@
                             <messageSpecFilter>*{Request,Response}.json</messageSpecFilter>
                             <templateDirectory>${project.basedir}/src/main/templates</templateDirectory>
-                            <templateNames>Kproxy/KrpcFilter.ftl</templateNames>
+                            <templateNames>Kproxy/TurboInvoker.ftl</templateNames>
```
This was a throwaway name, maybe it should be something like `SpecificMessageInvokeFilter`?

robobario commented on kroxylicious/src/test/java/io/kroxylicious/proxy/internal/FilterHandlerTest.java:
```diff
@@ -17,41 +17,117 @@
+        assertEquals(frame, propagated, "Expect it to be the frame that was sent");
+    }
+
+    private static class NoDeserialiseFilter implements ApiVersionsRequestFilter, RequestTargeter {
```
maybe instead of having separate interfaces for targeting all the generated Filter interfaces should have a method like: `shouldDeserializeApiVersionsRequest(short apiVersion)` defaulted to true. Then you could override the method for each type rather than having to switch on `apiKey`

k-wall replied:
I think I prefer this idea to the Targeters.  Alternatively someone expressed an idea about annotating the method with something that conveys the same information.

robobario commented:
From the [discussion](https://github.com/kroxylicious/design/discussions/10) I'm going to reduce the scope of this to adding an invoker and leave interface reworking for a separate PR

k-wall commented:
I'm really liking the direction this is heading. Thanks @robobario for pushing this forward.    +1 to the idea separating the inclusion of headers into the API to a future PR delivered after this one.

robobario commented:
Alright, I've reworked this to just implement just the Per-Message invoker changes to the Per-Message interfaces are removed. Also unsupported is implementing KrpcFilter directly, this could be done later after we discuss the API changes (like introducing RequestFilter, ResponseFilter interfaces instead)

tombentley commented on kroxylicious/src/main/templates/Kproxy/PerMessageInvoker.ftl:
```diff
@@ -0,0 +1,116 @@
+        if(filter instanceof KrpcFilter){
+            this.inner = (KrpcFilter) filter;
+        } else {
+            this.inner = null;
```
Is this case actually possible?

tombentley commented on kroxylicious/src/main/templates/Kproxy/PerMessageInvoker.ftl:
```diff
@@ -0,0 +1,116 @@
+import java.util.EnumSet;
+import java.util.Set;
+
+public class PerMessageInvoker implements KrpcFilter
```
I don't think the invocation of a filter needs to be conflated with the invoker being a filter itself. Or at least, the benefits of using inheritance aren't clear to me. Using inheritance like this, while convenient for the current code base, means we can't make KrpcFilter a `sealed` interface in the future without also exposing `PerMessageInvoker` in the `given` clause, but I definitely see `PerMessageInvoker` as an implementation detail.

robobario replied:
Okay, so I'm hearing we should reserve the KrpcFilter hierarchy for the filters. Maybe the Invoker should have a separate interface though so that we can build alternative invokers (like for the RequestFilter, ResponseFilter idea discussed).

tombentley commented on kroxylicious/src/main/templates/Kproxy/PerMessageInvoker.ftl:
```diff
@@ -0,0 +1,116 @@
+        return responseKeys.contains(apiKey) && (inner == null || inner.shouldDeserializeResponse(apiKey, apiVersion));
+    }
+
+    public static boolean implementsAnyMessageInterface(Object obj){
```
If we're keeping `KrpcFilter` as the base interface for the per-request and response interfaces then we should declare the parameter as a `KrpcFilter`: It's then the caller's job (specifically the `FilterChainFactory`) to do a downcast. It is anyway in a better position to provide a sensible exception message (relating it back to a specific config).

tombentley reviewed (commented):
I did a pass and left some comments. I've also got a branch with some experimental code, which takes a different route. TBH I don't think either approach is entirely satisfactory just yet. I'll try to push my branch so we can compare a contrast and figure out something which is satisfactory.

robobario commented on kroxylicious/src/main/java/io/kroxylicious/proxy/filter/FilterContributor.java:
```diff
@@ -12,5 +12,5 @@ public interface FilterContributor {
     Class<? extends FilterConfig> getConfigType(String shortName);
 
-    KrpcFilter getFilter(String shortName, ProxyConfig proxyConfig, FilterConfig filterConfig);
+    Object getFilter(String shortName, ProxyConfig proxyConfig, FilterConfig filterConfig);
```
So with that we would be able to `seal` the top interface but make all the generated sub-interfaces `non-sealed`, correct? I'm thinking about whether we could enable users to run a kroxy distribution but load custom filters pulled from a user-contributed jar (should probably be a separate discussion).

An empty marker interface sounds good for now.

robobario commented:
@tombentley I've taken those comments onboard and introduced a separate `FilterInvoker` interface for the invoker. It now operates on `KrpcFilter` rather than object. I also moved the construction of the invoker into the `KafkaProxyFrontendHandler`, so it creates the invokers and uses them to construct `FilterHandler`s and the `DecodePredicate`. So now the invoker and KrpcFilter do not share interfaces.

As you said it would be nice to move the construction up so we could log out which filter configuration we could not construct an invoker for, but we could add this later. Maybe the `FilterChainFactory` should produce an array of context object that contains the filter and invoker.

robobario commented:
/perf

robobario commented:
Closing in favour of #304
